libhl-crypto/src/signatures/mod.rs

```diff
@@ -0,0 +1,80 @@
+#[macro_use]
+mod macros;
+pub mod secp256k1;
+
+#[derive(Debug)]
+pub enum CryptoError {
+    /// Returned when trying to create an algorithm which does not exist.
+    NoSuchAlgorithm(String),
+    /// Returned when an error occurs during deserialization of a Private or
+    /// Public key from various formats.
+    ParseError(String),
+    /// Returned when an error occurs during the signing process.
+    SigningError(String),
+    /// Returned when an error occurs during key generation
+    KeyGenError(String),
+}
+
+/// A private key instance.
+/// The underlying content is dependent on implementation.
+pub trait PrivateKey<T: Sized + PublicKey> {
+    /// Returns the algorithm name used for this private key.
+    fn get_algorithm_name(&self) -> &str;
+    /// Sign a message
+    /// and return a byte array of the resulting signature.
+    /// # Arguments
+    /// * `message`- the message bytes to sign
+    /// # Returns
+    /// * `signature` - The signature in a binary array
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, CryptoError>;
+    /// Produce the public key for the given private key.
+    /// # Returns
+    /// * `public_key` - the public key for the given private key
+    fn get_public_key(&self) -> Result<T, CryptoError>;
+}
+
+/// A public key instance.
+/// The underlying content is dependent on implementation.
+pub trait PublicKey {
+    /// Returns the algorithm name used for this public key.
+    fn get_algorithm_name(&self) -> &str;
+    /// Verify a message with a signature
+    /// # Arguments
+    /// * `message` - the message bytes that were signed
+    /// * `signature` - the signature returned from *sign*
+    /// # Returns
+    /// * `boolean` - True if `message` this public key is associated with the signature,
+    ///               False otherwise
+    fn verify(&self, message: &[u8], signature: &[u8]) -> Result<bool, CryptoError>;
+}
+
+fn array_compare(a: &[u8], b: &[u8]) -> bool {
+    a.iter().enumerate().all(|(i, v)| *v == b[i])
+}
+
+fn bin2hex(b: &[u8]) -> String {
```

Comment:
I'm okay with that. I want to eliminate sodium oxide as a dependency because its no longer maintained and switch to another crate. I haven't found one that I like yet.

Reply:
The maitainers have changed but it looks like it is being maintained.